test/fixtures/es-module-loaders/mock-loader.mjs

```diff
@@ -0,0 +1,239 @@
+        return;
+      }
+      /**
+       * Fake modules created by `node:mock` get their meta.mock utility set
+       * to the corresponding value keyed off `mockedModules` and use this
+       * to setup their exports/listeners properly
+       */
+      if (context.url.startsWith('mock-facade:')) {
+        let [proto, version, encodedTargetURL] = context.url.split(':');
+        let decodedTargetURL = decodeURIComponent(encodedTargetURL);
+        if (mockedModules.has(decodedTargetURL)) {
+          meta.mock = mockedModules.get(decodedTargetURL);
+          return;
+        }
+      }
+      /**
+       * Ensure we still get things like `import.meta.url`
+       */
+      defaultImportMetaInitializer(meta, context);
+    });
+  };
+  return `(${insideAppContext})(getBuiltin, port, setImportMetaCallback)`
+}
+
+
+// Rewrites node: loading to mock-facade: so that it can be intercepted
+export function resolve(specifier, context, defaultResolve) {
+  if (specifier === 'node:mock') {
+    return {
+      url: specifier
+    };
+  }
+  doDrainPort();
+  const def = defaultResolve(specifier, context);
+  if (context.parentURL?.startsWith('mock-facade:')) {
+    // Do nothing, let it get the "real" module
+  } else if (mockedModuleExports.has(def.url)) {
+    return {
+      url: `mock-facade:${currentMockVersion}:${encodeURIComponent(def.url)}`
+    };
+  };
+  return {
+    url: `${def.url}`
+  };
+}
+
+export function load(url, context, defaultLoad) {
+  doDrainPort();
+  if (url === 'node:mock') {
+    /**
+     * Simply grab the import.meta.doMock to establish the communication
+     * channel with preloadCode
+     */
+    return {
+      source: 'export default import.meta.doMock',
+      format: 'module'
+    };
+  }
+  /**
+   * Mocked fake module, not going to be handled in default way so it
+   * generates the source text, then short circuits
+   */
+  if (url.startsWith('mock-facade:')) {
+    let [proto, version, encodedTargetURL] = url.split(':');
+    let ret = generateModule(mockedModuleExports.get(
+      decodeURIComponent(encodedTargetURL)
+    ));
+    return {
+      source: ret,
+      format: 'module'
+    };
+  }
+  return defaultLoad(url, context);
+}
+
+/**
+ * 
+ * @param {Array<string>} exports name of the exports of the module
+ * @returns {string}
+ */
+function generateModule(exports) {
+  let body = [
+    'export {};',
+    'let mapping = {__proto__: null};'
+  ];
+  for (const [i, name] of Object.entries(exports)) {
+    let key = JSON.stringify(name);
+    body.push(`var _${i} = import.meta.mock.namespace[${key}];`);
+    body.push(`Object.defineProperty(mapping, ${key}, { enumerable: true, set(v) {_${i} = v;}, get() {return _${i};} });`);
+    body.push(`export {_${i} as ${name}};`);
+  }
+  body.push(`import.meta.mock.listeners.push(${
+    () => {
+      for (var k in mapping) {
+        mapping[k] = import.meta.mock.namespace[k];
+      }
+    }
+  });`);
+  return body.join('\n');
+}
```

Comment:
239 lines for a fixture feels like a lot. Ideally we would have several smaller fixtures isolating chunks of the functionality to be tested, so that we could also have several smaller tests rather than one or a few big ones, so that if a test starts failing in the future it’s hopefully just a few (and pointing directly at the problem) rather than a big one. Maybe that’s impossible in this case, but something to consider.

Reply:
I don't see the real value in separating this into modules personally. It would just add complexity for a single test file.